stdlib/asm/crypto/dsa/rpo_falcon512.masm

```diff
@@ -218,182 +207,83 @@ export.load_h_s2_and_product.4
         push.M u32lt assert
         push.M u32lt assert
 
+        horner_eval_base
+
         hperm
     end
 
-    # 3) Load saved claimed hash of h and compare
+    # 4) Load PK, the saved claimed hash of h, and compare
     loc_loadw.0
     movup.4 assert_eq
     movup.3 assert_eq
     movup.2 assert_eq
     assert_eq
+    # => [C, ptr, alpha_inv, h(tau)_1, h(tau)_0, ...]
 
-    # 4) Load s2 (Due to the final norm test we do not need to range check the s2 coefficients)
-    padw padw
-    repeat.64
-        adv_pipe hperm
-    end
-
-    # 5) Load claimed h * s2 in Z_Q[x]
-    repeat.128
-        adv_pipe hperm
-    end
-
-    # 6) Return the challenge point and the incremented pointer
-    exec.rpo::squeeze_digest
-    drop drop
-    #=> [tau1, tau0, ptr + 512*4]
-end
-
-#! Checks that pi == h * s2 in Z_Q[x] by evaluating both sides at a random point.
-#! The procedure takes as input a pointer h_ptr to h. The other two polynomials
-#! are located at h_ptr + 512, for s2, and h_ptr + 1024, for pi. The procedure takes
-#! also a pointer zeros_ptr to a region of memory [zeros_ptr, zeros_ptr + 512*4)
-#! and a pointer tau_ptr to powers of the random point we are evaluating at stored
-#! as [a_i, b_i, x, x] where (a_i, b_i) := tau^i for i in [0, 1023].
-#! The procedure returns () if the check passes, otherwise it raises an exception
-#! related to an unsatisfied assertion.
-#!
-#! Input: [h_ptr, zeros_ptr, tau_ptr, ...]
-#! Output: [...]
-#!
-#! Cycles: 2504
-export.probabilistic_product.16
-    # 1) Save the pointers
-    push.0 movdn.3
-    loc_storew.0
-
-    # 2) Compute the evaluation of the h polynomial at the random challenge
-
-    # Accumulator to compute h(tau)
-    padw
-
-    # For mem_stream
+    # 5) Load s2 and evaluate at tau_inv (Due to the final norm test we do not need to range check the s2 coefficients)
```

Comment:
Do we explain somewhere in more detail why doing the final norm testing means no meed to range-check s2 coefficients?

Reply:
I spent some time investigating this and it seems that the argument justifying this comment no longer holds. I think this is a result of the recent changes or maybe even earlier.
Looking at the `norm_sq` procedure too, we are doing there a `u32lt` which would have undefined behavior if the input is not already checked to be a `u32`, so we should at least have a `u32` check on the coefficients of `s2`.
I have now added bound checks on the coefficients of `s2` similar to those on `h`.
This has unfortunately led to an increase in the number of cycles, the current cycle count is just shy of 60000 cycles. We can comeback to this to see if we can optimize these checks.

Reply:
Makes sense! Yeah - let's create an issue to figure out if we can do this range checking more efficiently somehow.

Reply:
Created one here https://github.com/0xPolygonMiden/miden-vm/issues/1672